New command: telepresence config view

thallgren opened the pull request:
## Description

This PR makes a few minor improvements to the Helm client config improvements, and adds a new `telepresence config view` command.

## Checklist

 - [x] I made sure to update `./CHANGELOG.md`.
 - [ ] I made sure to add any docs changes required for my change (including release notes).
 - [x] My change is adequately tested.
 - [ ] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.
 - [ ] I updated `TELEMETRY.md` if I added, changed, or removed a metric name.
 - [ ] Once my PR is ready to have integration tests ran, I posted the PR in #telepresence-dev in the datawire-oss slack so that the "ok to test" label can be applied.

Commit:
Add telepresence config view command.

Adds a new `telepresence config` command with one single `view`
subcommand. The subcommand takes one flag `--client-only` which limits
the output the current contents of the client's config.yml file. Without
a flag, the `config view` prints all configuration known to the daemons,
including everything that has been set from the traffic-manager.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Move command initialization into a file of its own.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Delete Config.UnmarshalYAML().

It's not strictly necessary (it's only there to output the "unknown key"
message), and it makes it impossible to inline the Config struct due to
this bug that never seem to get fixed:
https://github.com/go-yaml/yaml/issues/822

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add MarshalYAML and UnmarshalYAML to iputil.Subnet

So that it can be used when displaying the config.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Restore the Helm client dns and routing structures.

Those structs were replaced with what we normally use in the kubeconfig
extension, which wasn't so good because it breaks backward compatibility
and also mixes snake-case with camelCase names.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Update CHANGELOG.md

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Use yaml consistently for the traffic-manager-clients configmap.

The client parsed it as yaml, some comments indicated that it was yaml,
and now it is yaml everywhere. This also makes the configmap easier
to view and edit.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Watch traffic-manager-client configmap directly instead of mounting it.

Mounting a configmap and then using a file-watcher to get notified when
it changes is a very slow process (can take minutes). Watching the
configmap directly is however almost instant, and also a technique used
already by the agent-injector.

This commit replaces the mount of the configmap and the file-watcher
with a k8s watcher for the configmap.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Update tooling.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add GitHub action that pushes a release.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

raphaelreyna commented on pkg/client/cli/cmd_config.go:
```diff
@@ -0,0 +1,73 @@
+		}
+		cfg.Config = client.GetConfig(cmd.Context())
+		cfg.ClientFile = filepath.Join(cfgDir, client.ConfigFile)
+		// TODO: Also retrieve kubeconfig extension
```
we might want to actually do this before merging, seems incomplete to not include the kubeconfig extension.

thallgren replied:
Yep, I completely forgot about that one.

Commit:
Merge pull request #2873 from telepresenceio/thallgren/cliconfig-yaml

Helm chart client config improvments.

raphaelreyna reviewed (changes requested):
lgtm but I think its incomplete without showing the user the configs being pulled from the kubeconfig extension.

Commit:
Move things shared between the cli and userd out of userd.

The cli should never access things under userd and vice versa. They must
be thought of as separate executables.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add kubeconfig extension settings to config view client-only output.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
make generate

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix failing telepresence config view tests.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Merge pull request #2875 from telepresenceio/thallgren/gh-release

Add GitHub action that pushes a release.
